src/util/git-ref-resolver.js

```diff
@@ -0,0 +1,192 @@
+/* @flow */
+
+import type Config from '../config.js';
+import {removeSuffix} from './misc.js';
+
+const semver = require('semver');
+
+export type GitRefs = {
+  [name: string]: string,
+};
+export type DefaultBranch = {defaultBranch: true};
+export type ResolvedSha = {sha: string, ref: ?string};
+type Names = {tags: Array<string>, branches: Array<string>};
+
+export const isCommitSha = (target: string): boolean => Boolean(target) && /^[a-f0-9]{5,40}$/.test(target);
+
+const REF_TAG_PREFIX = 'refs/tags/';
+const REF_BRANCH_PREFIX = 'refs/heads/';
+
+// This regex is designed to match output from git of the style:
+//   ebeb6eafceb61dd08441ffe086c77eb472842494  refs/tags/v0.21.0
+// and extract the hash and ref name as capture groups
+const gitRefLineRegex = /^([a-fA-F0-9]+)\s+(refs\/(?:tags|heads)\/.*)$/;
+
+const refNameRegexp = /^refs\/(tags|heads)\/(.+)$/;
+
+class GitRefResolver {
+  constructor(config: Config, version: string, refs: GitRefs) {
+    this.config = config;
+    this.version = version;
+    this.refs = refs;
+  }
+
+  config: Config;
+  version: string;
+  refs: GitRefs;
+  names: ?Names;
+
+  async resolve(): Promise<DefaultBranch | ResolvedSha | false> {
+    const testFunctions = [
+      () => this.tryVersionAsGitCommit(),
```

Comment:
You can do
```js
GitRefResolver.testFunctions = [
    GitRefResolver.tryVersionAsFullRef,
    GitRefResolver.tryVersionAsTagName,
];
//...
async resolve(): Promise {
    for (const testFunction of testFunctions) {
        const result = await testFunction.call(this);
        if (result) {
            return result;
        }
    }
    return false;
}
```

Would save you from creating bound functions and an array each time `resolve` is called which could result in significant performance and memory savings.